### PR TITLE
lvalues should be checked for escapes, not rvalues

### PR DESCRIPTION
**Pull request by WalterBright**

### CODE REVIEW

**Comment by andralex** [comment]
any unittests that would fail without this?

**Comment by WalterBright** [comment]
There are no unittests for it. Right now, I'm checking to see if any of the test suite now fails.

**Comment by MartinNowak** [comment]
Thanks, both dub and vibe-d are failing (see https://ci.dawg.eu/job/projects/584/), because of the scope w/ lazy argument bug I send you.

```d
void log(Args)(lazy Args)
{
}

auto dg = (scope string[] vars) {
    log(vars);
};
```
`dmd -c -o- bug`

_This_ PR doesn't seem to introduce any new issues (though they could be hidden behind that other error).

_This_ PR does fix the other bug that I reported, here is the test case for reference.
```d
struct Cache
{
    ubyte[1] v;

    ubyte[] set(ubyte[1] v)
    {
        return this.v[] = v[];
    }
}
```
`dmd -c -o- bug`

**Comment by MartinNowak** [comment]
Fixing that revealed another bug in Sociomantic's ocean library (reproducible with the current state of dmd/scope and _this_ PR).
```d
class Argument
{
    string[] options;

    Argument restrict ( string[] options ... )
    {
        this.options = options;
        return this;
    }
}
```
dmd -c -o- bug.d

This is the 3rd and currently last one on my list, though I need to thoroughly retest.
Will also send this via mail in the same thread where the other 2 were reported.

**Comment by MartinNowak** [comment]
Also note that the bugs get more obscure (involving D variadic arguments) and I have to throw some 300KLOC at the compiler to find them, but the effort to reproduce them is fairly low and testing can be automated. My estimation, based on manual testing of projects, is that more than 50% of our regressions could have been catched by the project tester, so while it's still WIP, I'm really glad it's now running on every PR.

**Comment by MartinNowak** [comment]
Anything we're waiting on here?

**Comment by WalterBright** [comment]
> Anything we're waiting on here?

For this PR? Not that I know of.

**Comment by Geod24** [comment]
@MartinNowak : TBH the snippet you pointed to looks buggy as hell.
Typesafe variadics parameters are stack-allocated.

E.g. the following compiles:
```D
void main () @nogc
{
    auto test = foo("Foo", "Bar");
}

static string[] x;

string[] foo (string[] args...) @nogc
{
    x = args;
    return x;
}
```

While the following does not:
```D
void main () @nogc
{
    auto test = foo("Foo", "Bar");
}

static string[] x;

string[] foo (string[] args...) @nogc
{
    return args; // test.d(8): Error: escaping reference to variadic parameter args
}
```

**Comment by MartinNowak** [comment]
> For this PR? Not that I know of.

Then I'll integrate the test case so we can merge it.

**Comment by MartinNowak** [comment]
> @MartinNowak : TBH the snippet you pointed to looks buggy as hell.
> Typesafe variadics parameters are stack-allocated.

Wrong argumentation line, this might even be dead/unused code, but because it now suddenly fails to compile, it might block major projects or even whole companies.
The bottom line is fairly simple, we avoid to break code, no matter how broken it is, but we slowly purge people out of broken code using deprecations.

**Comment by MartinNowak** [comment]
The bug is already resolved, it's was a duplicate of [Issue 17029](https://github.com/dlang/dmd/pull/6363).

But please run the examples before reasoning, [here](http://forum.dlang.org/post/o39t7h$t4l$1@digitalmars.com) is the scope preview.
The error for https://github.com/dlang/dmd/pull/6365#issuecomment-269238347 was not about the variadic argument, but about returning this, which became scope when the variadic array was assigned to it's options fields, all of this happening without any switches. There is not even a `@safe` anywhere near that example.

File: test/compilable/interpret3.d
Comment (MartinNowak) [inline test/compilable/interpret3.d]: Why did you remove that test? It's surely escaping something, but the test has nothing to do with `-transition=safe/-dip1000`. It's also not `@safe` code.

Comment (WalterBright) [inline test/compilable/interpret3.d]: I removed it because it is a seriously broken piece of code, @safe or not. It is returning a pointer to its stack frame, and then dereferencing it. Allowing this kind of code in the name of backwards compatibility is a disservice to our users.

**Comment by MartinNowak** [comment]
Added the test case.

**Review by MartinNowak — CHANGES_REQUESTED** [review]

**Comment by WalterBright** [comment]
> The bottom line is fairly simple, we avoid to break code, no matter how broken it is, but we slowly purge people out of broken code using deprecations.

I agree with the sentiment, but this particular case goes too far. This is not an example of "working" code that we're breaking, it's a disaster to ship such code. It's a memory corruption bug waiting to strike if you look at it the wrong way. I doubt anyone is going to thank us for not pointing out the error.

File: test/compilable/interpret3.d
Comment (MartinNowak) [inline test/compilable/interpret3.d]: We had that discussion a couple of times already, and you did understood the argument the last time.
It doesn't matter whether the code is ugly, broken, incorrect, or reverse. You don't know it's role in D projects, it might never be executed, or it might just work out of chance, or it's a memory corruption that by chance causes no harms.
It might even affect a crucial system which can tolerate failures and retries until it succeeds.

Unless it's prohibitively complex to avoid breakage, you should not make random decisions on what is correct or wrong, because you're not the one to fix the million lines of D code out there.

> Allowing this kind of code in the name of backwards compatibility is a disservice to our users.

Nonsense, our service for D users is to transition to a safer language, so that such errors are detected and prevented, and doing that in a way that is predictable for people (e.g. a "6 month to fix that" deprecation).
Here you just break stuff completely unscheduled and with no idea about the consequences.

**Comment by MartinNowak** [comment]
> I agree with the sentiment, but this particular case goes too far. This is not an example of "working" code that we're breaking, it's a disaster to ship such code. It's a memory corruption bug waiting to strike if you look at it the wrong way. I doubt anyone is going to thank us for not pointing out the error.

The whole scope change is about pointing out such errors. But for whatever reason your arbitrarily selecting a random different date for that particular issue and don't tolerate any delay to fix it.
In effect dmd might refuse to compile code deeply nested in some random dependency.
So someone trying a D app, might get asked to fix some machine internals of e.g. vibe-d or the MySQL driver, even though she was never using that code.

**Comment by MartinNowak** [comment]
This is really not much of a reasoning or per-decision question. We were breaking code (and D projects) at a crazy rate, with hardly any plan. This has to stop, but whenever it comes to making such decision people use the same incorrect arguments to justify their actions.

- nobody should do that
- it was already broken
- that person will be thankful we pointed out their mistake
- that's ugly, let's help them

**Comment by WalterBright** [comment]
I suggest emailing the owners of that code that they have a memory corruption bug and need to fix it regardless. If I was the owner of such code I would appreciate that.

Such code can blow up at any time. Changing platforms, compiler switches, or compilers, all can trigger corruption. The corruption may be happening already and the user wasn't aware of it.

**Comment by WalterBright** [comment]
> The whole scope change is about pointing out such errors.

Not entirely. It also introduces new syntax which will be required even on correct code. That is gratuitous breakage, and so is properly behind a switch.

**Comment by WalterBright** [comment]
In any case, that debate is not relevant for this PR. What is blocking this PR?

**Comment by MartinNowak** [comment]
> I suggest emailing the owners of that code that they have a memory corruption bug and need to fix it regardless. If I was the owner of such code I would appreciate that.

Which doesn't work if you're on a schedule and need to get things done.
People can't always choose the time they have to update their compiler (e.g. b/c a new fix or feature is needed from an upstream library, and library now requires a newer compiler). Now if a compiler update randomly breaks other code, people are stuck.
Using deprecations instead of errors (even for obvious bugs) is the necessary tolerance to make an ecosystem with independently developed dependencies work. Without tolerance people are forced to fork and do everything on their own.

> In any case, that debate is not relevant for this PR. What is blocking this PR?

Indeed, I'd prefer to not have debates about fundamental process questions in PRs.
It is relevant though, b/c you've deliberately chosen to stop compiling certain code.

What would be the effort to continue supporting the above snippet?

**Comment by andralex** [comment]
Let's not now err on the side of being too conservative. All current C++ compilers change behavior on undefined code even across minor releases. May refuse to compile, may compile with different semantics, may even take advantage of it for optimization. I've seen each literally dozens of times. And we are talking about a language enjoying two orders of magnitude more use.

We are not in the business of keeping behavior on code with undefined behavior.

All - please let us not continue to argue. It is understood there are arguments and that reasonable people may disagree. This is final - let's move forward. Thanks!

Typed on the phone...

**Comment by MartinNowak** [comment]
> We are not in the business of keeping behavior on code with undefined behavior.

Well fine, no guaranteed backwards support for code with undefined low-level behavior. That's a clear policy we can communicate and work with.
What's problematic is the arbitrary choice and how easy similar argumentations are being used.

**Comment by MartinNowak** [comment]
The failure of CircleCI happens b/c the PR wasn't merged with master before testing, something addressed by https://github.com/dlang/dmd/pull/6361 on master.
The Project Tester failure comes from a not yet investigated linking error with the DerelictGLFW3 project, unrelated to this PR.

**Comment by MartinNowak** [comment]
Will just build a new dmd-scope preview, no known issues without the -transition=safe switch remaining.

**Comment by WalterBright** [comment]
Thank you, @MartinNowak

**Comment by andralex** [comment]
@MartinNowak thanks and yes good judgment is never in too much supply

**Comment by MartinNowak** [comment]
Can we add the decision on undefined code to http://dlang.org/deprecate or other appropriate places?

I wish we could leave this to common sense reasoning, but I've dealt with too many regressions that showed that to not scale. So there needs to be clearer line between the norm and exceptions thereof.

**Pull request merged**
